### ISSUE TITLE
Transparency missing in virtual keyboard on OpenDingux-beta

### ISSUE THREAD

**Issue by a-dekker** [szymor/vwordle#3]
Compiled with the OD-beta toolchain results in an absence of transparency in the virtual keyboard. This also means the selected letter is not highlighted and has to be selected blindly. Transparency does work correctly in the start screen itself. Any clue what goes wrong?

**Comment by szymor** [comment]
Well, hard to say until I reproduce the issue by myself, but my guess is that there is a problem in creating or loading bitmaps with alpha channel. Try to debug GameState::newColoredKeyboard() and surroundings to get more insight into the issue.

Is the virtual keyboard semitransparent and only letter highlight is not visible? Or the keyboard is opaque, you cannot see the highlight and guessed letters under the keyboard layer? The screenshot would be nice to be sure what we are talking about.

**Comment by a-dekker** [comment]
I was already looking/testing in GameState::newColoredKeyboard section for alpha settings regarding blending, but no success so far. This is how it looks right now:
![image](https://github.com/szymor/vwordle/assets/10128456/cdbaf212-823c-4140-a985-7d2ecfb469b8)

**Comment by szymor** [comment]
First you need to make sure the keyboard background (this black background with gradient at the edges) displays correctly. To do it, in file states.cpp please comment lines 425, 431 and 436. After that you will probably still see black rectangle with no gradient and letters. It is our starting point.
Before line 419 type:
`SDL_SetAlpha(keyboard_bg, SDL_SRCALPHA, 128)`
and please tell me the effect. You can also try to print out the surface parameters with printf:
`printf("R%d G%d B%d A%d SA%d bpp%d Bpp%d\n", keyboard_bg->format->Rmask, keyboard_bg->format->Gmask, keyboard_bg->format->Bmask, keyboard_bg->format->Amask, keyboard_bg->format->alpha, keyboard_bg->format->BitsPerPixel, keyboard_bg->format->BytesPerPixel)`
If you do above, also please share the results.

**Comment by a-dekker** [comment]
I already tested with commenting out line 249 (SDL_UnlockSurface(temp);) which leads to a correct transparency and letter highlight:
![screenshot003](https://github.com/szymor/vwordle/assets/10128456/2560c7a6-ad01-416c-9608-33b9a2c82737)


Applying your changes leads to the transparent background
![screenshot004](https://github.com/szymor/vwordle/assets/10128456/e1a7d6b3-9c26-458c-b4c8-a2791a5ec9a9)

The values that are printed are:
`R255 G65280 B16711680 A-16777216 SA128 bpp32 Bpp4`

**Comment by szymor** [comment]
Ok, so as far as I understand, the second image is after applying only my suggested changes. Please uncomment lines 425 and 431 (check_select and letter_select blits) to see if those two surface have the functioning alpha channel as well. If not, please add proper `SDL_SetAlpha(xxx, SDL_SRCALPHA, 128)` lines before blitting with proper surface names.

If that step is ok and we have proper alpha for all surfaces, we can do the same for line 436 (colored keyboard blitting).

If you have any issues at any step, please share the screenshot and related source code changes.

**Comment by a-dekker** [comment]
Yes, adding `SDL_SetAlpha(ckeyboard, SDL_SRCALPHA, 128);` before `SDL_BlitSurface(ckeyboard, NULL, screen, &dst);` does the trick!
![screenshot005](https://github.com/szymor/vwordle/assets/10128456/5fdc36db-04b6-4fb5-96dd-712534c64657)

**Comment by szymor** [comment]
I can see from the screenshot that there is still something wrong with alpha, could you call SDL_SetAlpha for keyboard_fg as well?

**Comment by a-dekker** [comment]
You are right, this way the letters look more smooth:
![screenshot006](https://github.com/szymor/vwordle/assets/10128456/44b32ea0-fe68-4b83-85c9-caed9f3c3c8b)

**Comment by szymor** [comment]
I can still see the issue with background. Are you sure you enabled SDL_SetAlpha for keyboard_bg? Could you please share your code changes?

**Comment by a-dekker** [comment]
I think I did:
```
  dst = tdst;
  SDL_SetAlpha(keyboard_bg, SDL_SRCALPHA, 128);
  SDL_BlitSurface(keyboard_bg, NULL, screen, &dst);
  dst = tdst;

```  
![screenshot007](https://github.com/szymor/vwordle/assets/10128456/900f5abb-ec0e-491a-8722-85a5d3b6d6dc)


If I compare it with your video (https://www.youtube.com/watch?v=Pp-OWjfy7ZQ) it looks a little darker, but perhaps that is brightness or screen related.

**Comment by a-dekker** [comment]
Increasing the brightness of the keyboard_bg image itself leads to more transparency
![screenshot008](https://github.com/szymor/vwordle/assets/10128456/f298569d-fa2b-423d-bad2-2f55d9e20147)

**Comment by szymor** [comment]
It surely is not about brightness, alpha mode for one of the surfaces (probably keyboard_bg) is not correct. If you do not care you do not need to debug it, but if you will, I could merge your changes to my repository. It is up to you.

In case you are interested in further debugging, you can print parameters of any surface with the following line:
`printf("R%08x G%08x B%08x A%08x SA%d bpp%d Bpp%d\n", keyboard_bg->format->Rmask, keyboard_bg->format->Gmask, keyboard_bg->format->Bmask, keyboard_bg->format->Amask, keyboard_bg->format->alpha, keyboard_bg->format->BitsPerPixel, keyboard_bg->format->BytesPerPixel)`
Just make sure to replace `keyboard_bg` with the surface you investigate. It would be good to know parameters of each surface involved in drawing the keyboard.

**Comment by szymor** [comment]
Please run the printf above for keyboard_fg and keyboard_bg, just to be sure.

**Comment by a-dekker** [comment]
added keyboard info at the start of the printf:
```
keyboard_bg: R000000ff G0000ff00 B00ff0000 Aff000000 SA128 bpp32 Bpp4
keyboard_fg: R000000ff G0000ff00 B00ff0000 Aff000000 SA128 bpp32 Bpp4
```

**Comment by a-dekker** [comment]
Is the lack of fade effect from the keyboard background to the game itself the issue you are referring to?

**Comment by szymor** [comment]
Yes, exactly. I do not understand why it worked at the beginning and now it does not.

**Comment by a-dekker** [comment]
Maybe I changed something unintentionally, let me check

**Comment by a-dekker** [comment]
Can't find anything unusual in the changes. 
The moment ckeyboard is shown, this fade effect goes away.
As soon as I comment the "SDL_BlitSurface(ckeyboard, NULL, screen, &dst)" block (near line 438), the screen looks like the second image I posted and the fade effect is back.

**Comment by szymor** [comment]
Yeah, the problem is that alpha value of ckeyboard should have no effect :) Alpha value in SDL_SetAlpha has no effect on RGBA surfaces. You can try to print surface parameters for ckeyboard, probably Amask will be 0x00 in this case. It should be fixed.

**Comment by szymor** [comment]
In line 190 of states.cpp please change SDL_HWSURFACE to 0 and share the results.

**Comment by a-dekker** [comment]
```
keyboard_bg: R000000ff G0000ff00 B00ff0000 Aff000000 SA1 bpp32 Bpp4
keyboard_fg: R000000ff G0000ff00 B00ff0000 Aff000000 SA128 bpp32 Bpp4
ckeyboard: R00ff0000 G0000ff00 B000000ff A00000000 SA128 bpp32 Bpp4
```
Replacing SDL_HWSURFACE with 0 leads to the second image: fade/blur effect, but no keys displayed.

**Comment by a-dekker** [comment]
I doubt if I will be able to fix this. And it is not really bothering me for now. Complements for this well designed game anyway.
Do you want me to create a PR for the changes so far, or do you have enough information? I have no idea of the changes will effect other build for other systems, probably not.

**Comment by szymor** [comment]
I do not want to merge an incomplete fix with unpredicted behaviour on other systems, so do not bother creating a PR. If I ever port the game to GCW0 or any compatible console, I will fix it on my own.

Thank you for the compliments :)
